File: Algoritmos/Trees/avl.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <assert.h>

struct node
{
    int val;
    int height;
    struct node *left;
    struct node *right;
};
typedef struct node node;

struct avl
{
    node *root;
    int size;
};
typedef struct avl avl;

// Return functions
int max(int n1, int n2)
{
    return n1 > n2 ? n1 : n2;
}

node *get_min(node *rt)
{
    if (rt->left == NULL)
        return rt;
    return get_min(rt->left);
}

node *delete_min(node *rt)
{
    if (rt->left == NULL)
        return rt->right;

    rt->left = delete_min(rt->left);
    return rt;
}

// Constructors
node* new_node(int val, node *left, node *right)
{
    node *n = (node *) malloc(sizeof(node));
    assert(n != NULL);

    n->val = val;
    n->left = left;
    n->right = right;
    n->height = -1;

    return n;
}

avl* create_tree()
{
    avl *t = (avl *) malloc(sizeof(avl));
    assert(t != NULL);

    t->root = NULL;
    t->size = 0;

    return t;
}

// Rotations
void update_height(node *rt)
{
    if (rt)
    {
        rt->height = 1;
        if (rt->left)
            rt->height += rt->left->height;
        if (rt->right)
            rt->height += rt->right->height;
    }
}

node *rightRotate(node *rt)
{
    node *l = rt->left;
    node *lr = l->right;

    l->right = rt;
    rt->left = lr;

    rt->height = 1 + max(rt->left ? rt->left->height : -1, rt->right ? rt->right->height : -1);

    l->height = 1 + max(l->left ? l->left->height : -1, l->right ? l->right->height : -1);
    
    update_height(rt);
    update_height(l);

    return l;
}

node *leftRotate(node *rt)
{
    node *r = rt->right;
    node *rl = r->left;

    r->left = rt;
    rt->right = rl;

    rt->height = 1 + max(rt->left ? rt->left->height : -1, rt->right ? rt->right->height : -1);

    r->height = 1 + max(r->left ? r->left->height : -1, r->right ? r->right->height : -1);

    update_height(rt);
    update_height(r);

    return r;
}

// Insertion
node* insert_help(node *rt, int val)
{
    if (rt == NULL)
        return new_node(val, NULL, NULL);
    
    if (rt->val > val)
        rt->left = insert_help(rt->left, val);
    else
        rt->right = insert_help(rt->right, val);
    
    rt->height = 1 + max(rt->left ? rt->left->height : -1, rt->right ? rt->right->height : -1);

    update_height(rt);

    int balance = (rt->left ? rt->left->height : -1) - (rt->right ? rt->right->height : -1);

    if ((balance < -1) && (val >= rt->right->val))
        return leftRotate(rt);
    if ((balance > 1) && (val < rt->left->val))
        return rightRotate(rt);
    if ((balance > 1) && (val >= rt->left->val))
    {
        rt->left = leftRotate(rt->left);
        return rightRotate(rt);
    }
    if ((balance < -1) && (val < rt->right->val))
    {
        rt->right = rightRotate(rt->right);
        return leftRotate(rt);
    }

    return rt;
}

void insert(avl *t, int val)
{
    t->root = insert_help(t->root, val);
    t->size++;
}

// Find
bool find_help(const node *rt, int val)
{
    if (rt == NULL)
        return false;
    if (rt->val > val)
        return find_help(rt->left, val);
    else if (rt->val == val)
        return true;
    
    return find_help(rt->right, val);
}

bool find(const avl *t, int val)
{
    return find_help(t->root, val);
}

// Remove
node *remove_help(node *rt, int val)
{
    if (rt == NULL)
        return NULL;

    if (rt->val > val)
        rt->left = remove_help(rt->left, val);
    else if (rt->val < val)
        rt->right = remove_help(rt->right, val);
    else
    {
        if (rt->left == NULL)
            return rt->right;
        else if (rt->right == NULL)
            return rt->left;
        else
        {
            node *temp = get_min(rt->right);
            rt->val = temp->val;
            rt->right = delete_min(rt->right);
        }
    }

    return rt;
}

void Remove(avl *t, int val)
{
    if (find_help(t->root, val))
    {
        t->root = remove_help(t->root, val);
        t->size--;
    }
}

// Traversies
void preorder(const node *rt)
{
    if (rt != NULL)
    {
        printf("%d ", rt->val);
        preorder(rt->left);
        preorder(rt->right);
    }
}
void inorder(const node *rt)
{
    if (rt != NULL)
    {
        inorder(rt->left);
        printf("%d ", rt->val);
        inorder(rt->right);
    }
}
void posorder(const node *rt)
{
    if (rt != NULL)
    {
        posorder(rt->left);
        posorder(rt->right);
        printf("%d ", rt->val);
    }
}

// Destructor
void posorder_destroy(node *rt)
{
    if (rt)
    {
        posorder_destroy(rt->left);
        posorder_destroy(rt->right);
        free(rt);
    }   
}

void destroy_avl(avl *t)
{
    posorder_destroy(t->root);
    free(t);
}

int main()
{
    avl *tree = create_tree();
    insert(tree, 10);
    insert(tree, 20);
    insert(tree, 30);
    insert(tree, 40);
    insert(tree, 50);
    insert(tree, 25);

    printf("Preorder traversal: ");
    preorder(tree->root);
    printf("\n");

    printf("Inorder traversal: ");
    inorder(tree->root);
    printf("\n");

    printf("Posorder traversal: ");
    posorder(tree->root);
    printf("\n");

    Remove(tree, 30);
    printf("After removing 30, inorder traversal: ");
    inorder(tree->root);
    printf("\n");

    destroy_avl(tree);
    return 0;
}

File: Algoritmos/Trees/Avl.go
package main

import ("fmt")

// Node structure
type node struct {
	elem int
	height int
	left *node
	right *node
}

func New_node(elem int, left *node, right *node) *node {
	return &node {elem, 0, left, right}
}

// AVL structure
type AVL struct {
	root *node
	size uint
}

func Create_AVL_tree() *AVL {
	return &AVL {nil, 0}
}

// Getters
func (tree *AVL) Root() *node {
	return tree.root
}

func (tree *AVL) Size() uint {
	return tree.size
}

func height(rt *node) int {
	if rt == nil {
		return -1
	}
	
	return 1 + max(height(rt.left), height(rt.right))
}

// Rotations
func right_rotate(rt *node) *node {
	l := rt.left
	lr := l.right

	l.right = rt
	rt.left = lr

	rt.height = 1 + max(height(rt.left), height(rt.right))
	l.height = 1 + max(height(l.left), height(l.right))

	return l
}

func left_rotate(rt *node) *node {
	r := rt.right
	rl := r.left

	r.left = rt
	rt.right = rl

	rt.height = 1 + max(height(rt.left), height(rt.right))
	r.height = 1 + max(height(r.left), height(r.right))

	return r
}

func get_balance(root *node) int {
	if root == nil {
		return 0
	}

	return (height(root.left) - height(root.right))
}

func insert_help(root *node, elem int) *node {
	if root == nil {
		return New_node(elem, nil, nil)
	}

	if (root.elem > elem) {
		root.left = insert_help(root.left, elem);
	} else {
		root.right = insert_help(root.right, elem);
	}	
	
	root.height = 1 + max(height(root.left), height(root.right))
	balance := get_balance(root);

	if (balance < -1) && (elem >= root.right.elem) {
		return left_rotate(root);
	}
	if (balance > 1) && (elem < root.left.elem) {
		return right_rotate(root);
	}
	if (balance > 1) && (elem >= root.left.elem) {
		root.left = left_rotate(root.left);
		return right_rotate(root);
	}
	if (balance < -1) && (elem < root.right.elem) {
		root.right = left_rotate(root.right);
		return left_rotate(root);
	}

	return root;
}

func (t *AVL) Insert(elem int) {
	t.root = insert_help(t.root, elem)
	t.size++
}

// Traverses
func (t *AVL) Preorder(root *node) {
	if root == nil {
		return
	}

	fmt.Printf("%v ", root.elem)
	t.Preorder(root.left)
	t.Preorder(root.right)
}

func (t *AVL) Inorder(root *node) {
	if root == nil {
		return
	}

	t.Inorder(root.left)
	fmt.Printf("%v ", root.elem)
	t.Inorder(root.right)
}

func (t *AVL) Posorder(root *node) {
	if root == nil {
		return
	}

	t.Posorder(root.left)
	t.Posorder(root.right)
	fmt.Printf("%v ", root.elem)
}
